coderd/provisionerdserver/provisionerdserver.go

```diff
@@ -1190,7 +1190,12 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
 						// have a shutdown signal we can listen to.
 						<-wait
 						if err := s.Pubsub.Publish(codersdk.WorkspaceNotifyChannel(workspaceBuild.WorkspaceID), []byte{}); err != nil {
-							s.Logger.Error(ctx, "workspace notification after agent timeout failed",
+							// If the publish failed due to the context being canceled, there's nothing more for us
+							// to do here.
```

Comment:
Does this mean the provisioner is shutting down? Should we mention that?

Reply:
From my reading, it means the `coderd` API that talks to provisioners is shutting down, as provisioner daemons have no access to the database pubsub directly.

Reply:
This is the error that `Publish` returns, and if it's `context.Canceled`, it must mean the pubsub itself is shutting down (note that `Publish` does not take a context).

This feels very brittle.  Here and now, we know that `context.Canceled` means that the pubsub is shutting down, and here and now, we're pretty sure that means that Coderd itself is shutting down, but there is no contract with `Pubsub` that it will return `context.Canceled` on shutdown, and there is no contract that Pubsub only gets shut down when everything is being shut down.

What I did on #10170 was just ignore errors, but it seems the issue is more widespread, so I was probably wrong to take the easy way out.

What we should really do is pass a context to the provisionerd server, and then 

```
select {
    case <-s.ctx.Done():
        return
    case <-wait:
        // process the timeout
}
```

The context we pass to provisionerdserver should be the Coderd lifetime context.

Reply:
@spikecurtis Updated. I think I threaded it through correctly but you know these streets better than I.